### PR TITLE
Ptx 11665

### PR DESCRIPTION
**Pull request by gauraa2**
PDS System test automation: 
    Summary - Rolling reboot with data service running.
    Ref doc: https://portworx.atlassian.net/wiki/spaces/TA/pages/2505769036/PDS+System+test+automation
    Jira ticket- https://portworx.atlassian.net/browse/PTX-11665

### CODE REVIEW

File: drivers/pds/api/backupCredential.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+	if res.StatusCode != status.StatusOK {
+		log.Errorf("Error when calling `ApiTenantsIdBackupCredentialsPost``: %v\n", err)
+		log.Errorf("Full HTTP response: %v\n", res)
+	}
+	return backupModel, err
+
+}
+
+// CreateS3BackupCredential func
+func (backupCredential *BackupCredential) CreateS3BackupCredential(tenantID string, name string, accessKey string, endpoint string, secretKey string) (*pds.ModelsBackupCredentials, error) {
```
Comment (lsrinivas-pure) [inline drivers/pds/api/backupCredential.go:65]: if these params are common to all functions, can we set them as part of struct itself

File: drivers/pds/api/backupPolicy.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	if res.StatusCode != status.StatusOK {
+		log.Errorf("Error when calling `ApiTenantsIdBackupPoliciesPost``: %v\n", err)
+		log.Error("Full HTTP response: %v\n", res)
+	}
+	return backupPolicyModel, err
+
+}
+
+// UpdateBackupPolicy func
+func (backupPolicy *BackupPolicy) UpdateBackupPolicy(backupCredsID string, name string, retentionCount int32, scheduleCronExpression string, backupType string) (*pds.ModelsBackupPolicy, error) {
```
Comment (lsrinivas-pure) [inline drivers/pds/api/backupPolicy.go:63]: same here

File: drivers/pds/lib/controlplane.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		"client_id":     clientID,
+		"client_secret": clientSecret,
+		"username":      username,
+		"password":      password,
+	})
+
+	requestBody := bytes.NewBuffer(postBody)
+	resp, err := http.Post(url, "application/json", requestBody)
+	if err != nil {
+		log.Fatalf("An Error Occured %v", err)
```
Comment (lsrinivas-pure) [inline drivers/pds/lib/controlplane.go:50]: retun the error and fail it in the test

File: drivers/pds/lib/controlplane.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+	saClient := cp.components.ServiceAccount
+	serviceAccounts, _ := saClient.ListServiceAccounts(tenantID)
+	var agentWriterID string
+	for _, sa := range serviceAccounts {
+		if sa.GetName() == "Default-AgentWriter" {
+			agentWriterID = sa.GetId()
+		}
+	}
+	token, _ := saClient.GetServiceAccountToken(agentWriterID)
```
Comment (lsrinivas-pure) [inline drivers/pds/lib/controlplane.go:80]: retun err in cae of error

File: drivers/pds/lib/targetcluster.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+	apiEndpoint := fmt.Sprintf(controlPlaneURL + "/api")
+	log.Infof("Installing PDS ( helm version -  %v)", helmChartversion)
+	if strings.EqualFold(clusterType, "ocp") {
+		cmd = fmt.Sprintf("helm install --create-namespace --namespace=%s pds pds-target --repo=https://portworx.github.io/pds-charts --version=%s --set platform=ocp --set tenantID=%s "+
+			"--set bearerToken=%s --set apiEndpoint=%s --kubeconfig %s", PDSNamespace, helmChartversion, tenantID, bearerToken, apiEndpoint, targetCluster.kubeconfig)
+	} else {
+		cmd = fmt.Sprintf("helm install --create-namespace --namespace=%s pds pds-target --repo=https://portworx.github.io/pds-charts --version=%s --set tenantID=%s "+
+			"--set bearerToken=%s --set apiEndpoint=%s --kubeconfig %s", PDSNamespace, helmChartversion, tenantID, bearerToken, apiEndpoint, targetCluster.kubeconfig)
+	}
+	log.Warn(cmd)
```
Comment (lsrinivas-pure) [inline drivers/pds/lib/targetcluster.go:67]: why warn?

File: drivers/pds/lib/targetcluster.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+	time.Sleep(10 * time.Second)
```
Comment (lsrinivas-pure) [inline drivers/pds/lib/targetcluster.go:68]: why do u require sleep ?

File: drivers/pds/lib/targetcluster.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+	output, _, err := osutils.ExecShell(cmd)
+	if err != nil {
+		log.Warn("Kindly remove the PDS chart properly and retry if that helps(or slack us for more details). CMD>> helm uninstall  pds --namespace pds-system --kubeconfig $KUBECONFIG")
+		log.Panic(err)
```
Comment (lsrinivas-pure) [inline drivers/pds/lib/targetcluster.go:72]: after panic return wont get executed i believe

File: drivers/pds/lib/targetcluster.go [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+		return err
+	}
+	log.Infof("Terminal output -> %v", output)
+	err = targetCluster.ValidatePDSComponents()
+	if err != nil {
+
```
Comment (lsrinivas-pure) [inline drivers/pds/lib/targetcluster.go:78]: return err here

File: drivers/pds/lib/targetcluster.go [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return nsObj.(*corev1.Namespace), nil
+}
+
+// GetClusterID of target cluster.
+func (targetCluster *TargetCluster) GetClusterID() (string, error) {
+	log.Infof("Fetch Cluster id ")
+	cmd := fmt.Sprintf("kubectl get ns kube-system -o jsonpath={.metadata.uid} --kubeconfig %s", targetCluster.kubeconfig)
```
Comment (lsrinivas-pure) [inline drivers/pds/lib/targetcluster.go:165]: can u check if we already have get namespace util

File: pkg/osutils/osutils.go [hunk 10]
```diff
@@ -123,3 +126,66 @@ func Kubectl(arguments []string) error {
 
 	return nil
 }
+
+// Exec Function to execute local command
+func Exec(command string, arg ...string) (string, error) {
+	out, err := exec.Command(command, arg...).CombinedOutput()
+	outStr := string(out)
+	return outStr, err
+}
```
Comment (lsrinivas-pure) [inline pkg/osutils/osutils.go:135]: can we use existing RunCommand( ssh.go) or execpod functions

File: tests/pds/pds_system_test.go [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+				dataServiceNameVersionMap               = make(map[string][]string)
+				dataServiceIDImagesMap                  = make(map[string]string)
+				dataServiceNameDefaultAppConfigMap      = make(map[string]string)
+				deployementIDNameMap                    = make(map[string]string)
+				namespaceNameIDMap                      = make(map[string]string)
+			)
+
+			clusterID, err := targetCluster.GetClusterID()
+			if err != nil {
+				log.Panicf("Unable to fetch the cluster Id")
```
Comment (lsrinivas-pure) [inline tests/pds/pds_system_test.go:160]: use Expect here

File: tests/pds/pds_system_test.go [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+			}
+
+			log.Info("Get the Target cluster details")
+			targetClusters, _ := deploymentTargetComponent.ListDeploymentTargetsBelongsToTenant(tenant.GetId())
```
Comment (lsrinivas-pure) [inline tests/pds/pds_system_test.go:164]: handle the error

File: tests/pds/pds_system_test.go [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+			for i := 0; i < len(targetClusters); i++ {
+				if targetClusters[i].GetClusterId() == clusterID {
+					deploymentTargetID = targetClusters[i].GetId()
+					log.Infof("Cluster ID: %v, Name: %v,Status: %v", targetClusters[i].GetClusterId(), targetClusters[i].GetName(), targetClusters[i].GetStatus())
+				}
+			}
+
+			log.Infof("Get the available namespaces in the Cluster having Id: %v", clusterID)
+			namespaces, _ := nsComponent.ListNamespaces(deploymentTargetID)
+			for i := 0; i < len(namespaces); i++ {
```
Comment (lsrinivas-pure) [inline tests/pds/pds_system_test.go:174]: same as above

File: tests/pds/pds_system_test.go [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+				if namespaces[i].GetStatus() == "available" {
+					namespaceNameIDMap[namespaces[i].GetName()] = namespaces[i].GetId()
+					log.Infof("Available namespace - Name: %v , Id: %v , Status: %v", namespaces[i].GetName(), namespaces[i].GetId(), namespaces[i].GetStatus())
+				}
+			}
+			log.Infof("Get the storage template")
+			storageTemplates, _ := storagetemplateComponent.ListTemplates(tenant.GetId())
```
Comment (lsrinivas-pure) [inline tests/pds/pds_system_test.go:181]: same as above

File: tests/pds/pds_system_test.go [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+
+			log.Infof("Get the Versions.")
+			for key := range dataServiceNameIDMap {
+				versions, _ := versionComponent.ListDataServiceVersions(dataServiceNameIDMap[key])
+				for i := 0; i < len(versions); i++ {
+					dataServiceNameVersionMap[key] = append(dataServiceNameVersionMap[key], versions[i].GetId())
+				}
+			}
+
+			log.Infof("Get the Versions.")
```
Comment (lsrinivas-pure) [inline tests/pds/pds_system_test.go:221]: mention whixh versions

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]
check the compilation errors

File: drivers/pds/api/backupCredential.go (hunk 1 above)
Comment (gauraa2) [reply drivers/pds/api/backupCredential.go:65]: These params are not common to all functions hence, didn't added as part of the struct.

File: drivers/pds/api/backupPolicy.go (hunk 2 above)
Comment (gauraa2) [reply drivers/pds/api/backupPolicy.go:63]: These params are not common to all functions hence, didn't added as part of the struct.

File: drivers/pds/lib/controlplane.go (hunk 3 above)
Comment (gauraa2) [reply drivers/pds/lib/controlplane.go]: Done

File: drivers/pds/lib/controlplane.go (hunk 4 above)
Comment (gauraa2) [reply drivers/pds/lib/controlplane.go]: Done

File: drivers/pds/lib/targetcluster.go (hunk 5 above)
Comment (gauraa2) [reply drivers/pds/lib/targetcluster.go]: Updated as Info.

File: tests/pds/pds_system_test.go (hunk 11 above)
Comment (gauraa2) [reply tests/pds/pds_system_test.go]: Done.

File: tests/pds/pds_system_test.go (hunk 12 above)
Comment (gauraa2) [reply tests/pds/pds_system_test.go]: Done

File: tests/pds/pds_system_test.go (hunk 13 above)
Comment (gauraa2) [reply tests/pds/pds_system_test.go:180]: Done

File: tests/pds/pds_system_test.go (hunk 14 above)
Comment (gauraa2) [reply tests/pds/pds_system_test.go:188]: Done

File: pkg/osutils/osutils.go (hunk 10 above)
Comment (gauraa2) [reply pkg/osutils/osutils.go:135]: This is supposed to be run on the host VM / Jenkins VM where helm / kubectl is already installed and doesn't require to execute on any node. Hence not utilizing the RunCommand which require node object as param.

File: tests/pds/pds_system_test.go (hunk 14 above)
Comment (lsrinivas-pure) [reply tests/pds/pds_system_test.go:188]: what is the _ here?

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]
I see checks have failed. can you please have a look

**Comment by gauraa2** [comment]
Updated : https://github.com/portworx/torpedo/pull/963

**Pull request closed without merging**